src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs

```diff
@@ -5349,5 +5349,222 @@ static void Main()
             string expectedOutput = @"1";
             CompileAndVerify(source, expectedOutput: expectedOutput);
         }
+
+        [Fact]
+        public void TestNullCoalesce_NullableWithDefault_Optimization()
+        {
+            var source = @"
+class Program
+{
+    struct S
+    {
+        System.DateTime a;
+        System.Guid b;
+    }
+
+    static int CoalesceInt32(int? x)
+    {
+        return x ?? 0;
+    }
+
+    static T CoalesceGeneric<T>(T? x) where T : struct
+    {
+        return x ?? default(T);
+    }
+
+    static (bool a, System.TimeSpan b) CoalesceTuple((bool a, System.TimeSpan b)? x)
+    {
+        return x ?? default((bool a, System.TimeSpan b));
+    }
+
+    static S CoalesceUserStruct(S? s)
+    {
+        return s ?? default(S);
+    }
+
+    static S CoalesceStructWithImplicitConstructor(S? x)
+    {
+        return x ?? new S();
+    }
+}";
+            var comp = CompileAndVerify(source);
```

Comment:
>CompileAndVerify(source) [](start = 23, length = 24)

Besides verifying IL, all tests for success cases should execute the code and observe expected behavior for null and not null input. Please make necessary adjustments to this and other similar tests.

Reply:
Done